### PR TITLE
Make preload query to prepared statements

### PR DESCRIPTION
**Pull request by kamipo**
Currently preload query cannot be prepared statements even if
`prepared_statements: true`. This commit fixes the issue.

### CODE REVIEW

**Comment by sgrif** [comment]
Instead of handling it there, can we fix the predicate builder to use prepared statements any time it's passed an array of one item?

**Comment by kamipo** [comment]
It is possible, but more complicated. To decouple building Arel ASTs from uniqueness validator, the predicate builder needs to support `array` and `range` types. #26072 
It seems the handling of arrays in the predicate builder becomes complicated.

**Comment by kamipo** [comment]
I tried to fix the predicate builder to handle an array of one item. Now the predicate builder uses prepared statements any time when passed an array of one item. Thanks!

**Comment by sgrif** [comment]
Can you rebase?

**Comment by kamipo** [comment]
Rebased!

**Comment by matthewd** [comment]
Does this conflict with my desire to use prepared statements for all array predicates, by binding the array as a single value, on array-supporting RDBMSes? :confused:

File: activerecord/lib/active_record/relation/predicate_builder.rb [hunk 1]
```diff
@@ -95,6 +95,11 @@ def create_binds_for_hash(attributes)
             next
           when value.is_a?(Relation)
             binds += value.bound_attributes
+          when value.is_a?(Array) && !table.type(column_name).respond_to?(:subtype)
+            if value.size == 1 && can_be_bound?(column_name, value.first)
+              result[column_name] = Arel::Nodes::BindParam.new
+              binds << build_bind_param(column_name, value.first)
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: If RDBMSes supports an array type and the column is an array type, it is bound as an array value.

```ruby
  PgArray.find_by(tags: ["black"])
  # => SELECT  "pg_arrays".* FROM "pg_arrays" WHERE "pg_arrays"."tags" = $1 LIMIT $2  [["tags", "{black}"], ["LIMIT", 1]]
```

**Comment by matthewd** [comment]
@kamipo I mean where we currently do:

```ruby
User.where(id: [1, 2])
# => SELECT ... WHERE id IN (1, 2)
```

I'd like to instead produce:

```ruby
User.where(id: [1, 2])
# => SELECT ... WHERE id = ANY($1)   ["{1,2}"]
```

.. as this allows us to prepare a statement once, and not need a new one for different array lengths.

It's currently just a future thought, but I wondered if introducing a condition here might make it harder to do the above (which obviously must be adapter specific) later.

**Comment by kamipo** [comment]
I think it is easy to support the above later by the condition.

```diff
--- a/activerecord/lib/active_record/relation/predicate_builder.rb
+++ b/activerecord/lib/active_record/relation/predicate_builder.rb
@@ -96,7 +96,7 @@ def create_binds_for_hash(attributes)
           when value.is_a?(Relation)
             binds += value.bound_attributes
           when value.is_a?(Array) && !table.type(column_name).respond_to?(:subtype)
-            if value.size == 1 && can_be_bound?(column_name, value.first)
+            if !supports_array_types? && value.size == 1 && can_be_bound?(column_name, value.first)
               result[column_name] = Arel::Nodes::BindParam.new
               binds << build_bind_param(column_name, value.first)
             end
```

File: activerecord/lib/active_record/relation/predicate_builder.rb [hunk 2]
```diff
@@ -95,6 +95,11 @@ def create_binds_for_hash(attributes)
             next
           when value.is_a?(Relation)
             binds += value.bound_attributes
+          when value.is_a?(Array) && !table.type(column_name).respond_to?(:subtype)
```
Comment (sgrif) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: We should not be relying on the internals of type objects here.

Comment (sgrif) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: This change seems unrelated to the rest of the pull request.

Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: If removed the type condition, `test_where_by_attribute_with_array` fails.
https://github.com/rails/rails/blob/master/activerecord/test/cases/adapters/postgresql/array_test.rb#L293-L297

```ruby
  def test_where_by_attribute_with_array
    tags = ["black", "blue"]
    record = PgArray.create!(tags: tags)
    assert_equal record, PgArray.where(tags: tags).take
  end
```

```
# correct
SELECT  "pg_arrays".* FROM "pg_arrays" WHERE "pg_arrays"."tags" = $1 LIMIT $2  [["tags", "{black,blue}"], ["LIMIT", 1]]

# removed type condition
SELECT  "pg_arrays".* FROM "pg_arrays" WHERE "pg_arrays"."tags" IN ('black', 'blue') LIMIT $1  [["LIMIT", 1]]
```

**Comment by sgrif** [comment]
Wait when was that added? That isn't supposed to work in the first place.

File: activerecord/lib/active_record/relation/predicate_builder.rb [hunk 3]
```diff
@@ -112,6 +112,12 @@ def create_binds_for_hash(attributes)
               binds.concat(bvs)
               attrs
             end
+          when value.is_a?(Array) && !table.type(column_name).respond_to?(:subtype)
```
Comment (sgrif) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: I'd really prefer not to add more code which relies on type internals in this way. I think the code below this which does the same is a major smell. There's a greater problem with the structure here that we should address first.

Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: I agree that there's a problem with the structure.
How about the alternative #29033?
#29033 is a simple one line fix and we already have a similar code (for keeping existing expectation: https://github.com/rails/rails/blob/a6d6e90fa90eedbdb6d3cda679aa4b47ddfa7315/activerecord/lib/active_record/relation/predicate_builder/polymorphic_array_value.rb#L13).
I'd like to add the expectation that typical generated queries will be preparable even if association queries.

Comment (sgrif) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: Yeah I like #29033. Should we close this one?

Comment (kamipo) [inline activerecord/lib/active_record/relation/predicate_builder.rb]: Yeah, I'm closing this for now.

**Comment by kamipo** [comment]
Closing in favor of #29033.
